ores/scoring_context.py

```diff
@@ -0,0 +1,224 @@
+                self.process_score(model_name, dependency_cache=root_cache)
+
+            # Essentially free
+            if include_features:
+                base_feature_map = self.solve_base_feature_map(
+                    model_name, dependency_cache=root_cache)
+                model_scores[model_name]['features'] = base_feature_map
+
+        return model_scores
+
+    def solve_features(self, model_name, dependency_cache=None):
+        """
+        """
+        features = self[model_name].features
+        return list(self.extractor.solve(features, cache=dependency_cache))
+
+    def solve_base_feature_map(self, model_name, dependency_cache=None):
+        """
+        """
+        features = list(trim(self[model_name].features))
+        feature_values = self.extractor.solve(features, cache=dependency_cache)
+        return {str(f): v
+                for f, v in zip(features, feature_values)}
+
+    def process_score(self, model_name, dependency_cache=None):
+        """
+        """
+        version = self[model_name].version
+
+        start = time.time()
+        feature_values = list(self.solve_features(
+            model_name, dependency_cache))
+        logger.debug("Extracted features for {0}:{1}:{2} in {3} secs"
+                     .format(self.name, model_name, version,
+                             time.time() - start))
+
+        start = time.time()
+        score = self[model_name].score(feature_values)
+        logger.debug("Scored features for {0}:{1}:{2} in {3} secs"
+                     .format(self.name, model_name, version,
+                             time.time() - start))
+
+        return score
+
+    def _generate_root_datasources(self, model_names):
+        for model_name in model_names:
+            for dependency in dependencies.dig(self.model_features(model_name)):
+                if isinstance(dependency, Datasource):
+                    yield dependency
+
+    def extract_root_dependency_caches(
+            self, model_names, rev_ids, injection_caches=None):
+        """
+        Extracts a mapping of root datasources capable of generating the
+        features needed for a particular set of models.
+
+        :Parameters:
+            model_names : `list` ( `str` )
+                The names of a :class:`~revscoring.ScorerModel` to
+                extract the roots dependencies for
+        """
+        # Make a copy of dependency_caches
+        root_caches = {}
+        for rev_id in rev_ids:
+            injection_cache = injection_caches.get(rev_id) \
+                              if injection_caches is not None else {}
+            root_caches[rev_id] = {key: value
+                                   for key, value in injection_cache.items()}
+
+        # Find our root datasources
+        root_datasources = \
+            list(set(self._generate_root_datasources(model_names)))
+
+        # Extract the root datasource and update the root_caches in-place
```

Comment:
the side-effect should be more explicitly described through something other than a comment in the caller's code